core/client/hammer/hammer.go

```diff
@@ -76,125 +96,187 @@ func New(ctx context.Context, dial DialFunc, callOptions CallOptions,
 	}, nil
 }
 
-// Run runs operationCount with up to maxWorkers over ramp duration.
-func (h *Hammer) Run(ctx context.Context, operationCount, maxWorkers int, ramp time.Duration) {
-	jobs := gen(ctx, operationCount)
-	outputs := h.launchWorkers(ctx, jobs, maxWorkers, ramp)
-	h.printStats(1*time.Second, outputs)
-}
+// Run runs operationCount with up to maxWorkers.
+// workers should roughly be (goal QPS / timeout seconds)
+func (h *Hammer) Run(ctx context.Context, numWorkers int, c Config) error {
+	workers, err := h.newWorkers(numWorkers)
+	if err != nil {
+		return err
+	}
 
-// launchWorkers gradually adds workers to the worker pool over ramp up to workers.
-func (h *Hammer) launchWorkers(ctx context.Context, jobs <-chan int, workers int, ramp time.Duration) <-chan error {
-	outputs := make(chan error)
+	if ok := c.TestTypes["batch"]; ok {
+		// Batch Write users
+		log.Printf("Batch Write")
+		requests := requestGenerator(ctx, c.BatchWriteQPS, c.BatchWriteSize, c.BatchWriteCount, c.Duration)
+		handlers := make([]ReqHandler, 0, len(workers))
+		for _, w := range workers {
+			handlers = append(handlers, w.writeOp)
+		}
+		log.Printf("workers: %v", len(handlers))
+		executeRequests(ctx, requests, handlers)
+		fmt.Print("\n")
+	}
 
-	go func() {
-		rampTicker := time.NewTicker(ramp / time.Duration(workers))
-		var wg sync.WaitGroup
-		for h.workers < workers {
-			h.workers++
-			wg.Add(1)
-			go func() {
-				defer wg.Done()
-				h.worker(ctx, jobs, outputs)
-			}()
-			<-rampTicker.C
+	if ok := c.TestTypes["write"]; ok {
+		// Write users
+		log.Printf("User Write")
+		requests := requestGenerator(ctx, c.WriteQPS, 1, c.WriteCount, c.Duration)
+		handlers := make([]ReqHandler, 0, len(workers))
+		for _, w := range workers {
+			handlers = append(handlers, w.writeOp)
 		}
-		rampTicker.Stop()
+		executeRequests(ctx, requests, handlers)
+		fmt.Print("\n")
+	}
 
-		wg.Wait()
-		close(outputs)
-	}()
+	if ok := c.TestTypes["read"]; ok {
+		// Read users
+		log.Printf("User Read")
+		requests := requestGenerator(ctx, c.ReadQPS, c.ReadPageSize, c.ReadCount, c.Duration)
+		handlers := make([]ReqHandler, 0, len(workers))
+		for _, w := range workers {
+			handlers = append(handlers, w.readOp)
+		}
+		executeRequests(ctx, requests, handlers)
+		fmt.Print("\n")
+	}
 
-	return outputs
+	if ok := c.TestTypes["audit"]; ok {
+		// History
+		log.Printf("User Audit History")
+		requests := requestGenerator(ctx, c.HistoryQPS, c.HistoryPageSize, c.HistoryCount, c.Duration)
+		handlers := make([]ReqHandler, 0, len(workers))
+		for _, w := range workers {
+			handlers = append(handlers, w.historyOp)
+		}
+		executeRequests(ctx, requests, handlers)
+		fmt.Print("\n")
+	}
+
+	return nil
 }
 
-// gen produces a channel of integers from 0 up to count-1 and then closes.
-func gen(ctx context.Context, count int) <-chan int {
-	jobs := make(chan int)
+func requestGenerator(ctx context.Context, qps, batch, count int, duration time.Duration) <-chan request {
```

Comment:
Maybe `gen(erate)Requests`? Otherwise `request` can be read as a verb, which is slightly misleading.